presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java

```diff
@@ -13,202 +13,300 @@
+                }
+            }
+            catch (Throwable t) {
+                throw internalError(t);
+            }
+        }
+        if (hasNull) {
+            return null;
+        }
+        return false;
     }
 
-    @SqlNullable
-    @TypeParameter("E")
+    @TypeParameter("T")
+    @TypeParameterSpecialization(name = "T", nativeContainerType = long.class)
     @SqlType(StandardTypes.BOOLEAN)
-    public Boolean arraysOverlap(
-            @OperatorDependency(operator = LESS_THAN, argumentTypes = {"E", "E"}) MethodHandle lessThanFunction,
-            @TypeParameter("E") Type type,
-            @SqlType("array(E)") Block leftArray,
-            @SqlType("array(E)") Block rightArray)
+    @SqlNullable
+    public static Boolean arraysOverlapLong(
+            @OperatorDependency(operator = EQUAL, argumentTypes = {"T", "T"}) MethodHandle equals,
+            @TypeParameter("T") Type elementType,
+            @SqlType("array(T)") Block leftArray,
+            @SqlType("array(T)") Block rightArray)
     {
-        int leftPositionCount = leftArray.getPositionCount();
-        int rightPositionCount = rightArray.getPositionCount();
-
-        if (leftPositionCount == 0 || rightPositionCount == 0) {
+        if (leftArray.getPositionCount() == 0 || rightArray.getPositionCount() == 0) {
             return false;
         }
 
-        if (leftPositions == null || leftPositions.length < leftPositionCount) {
-            leftPositions = new int[leftPositionCount * 2];
+        if (Math.max(rightArray.getPositionCount(), leftArray.getPositionCount()) >= NESTED_LOOP_THRESHOLD) {
+            return arraysOverlapSetBased(elementType, leftArray, rightArray);
         }
-
-        if (rightPositions == null || rightPositions.length < rightPositionCount) {
-            rightPositions = new int[rightPositionCount * 2];
+        boolean hasNull = false;
+        for (int i = 0; i < leftArray.getPositionCount(); i++) {
+            if (leftArray.isNull(i)) {
+                hasNull = true;
+                continue;
+            }
+            try {
+                Long leftValue = elementType.getLong(leftArray, i);
+                for (int j = 0; j < rightArray.getPositionCount(); j++) {
+                    if (rightArray.isNull(j)) {
+                        hasNull = true;
+                        continue;
+                    }
+                    if ((Boolean) equals.invokeExact(elementType.getLong(rightArray, j), leftValue.longValue())) {
+                        return true;
+                    }
+                }
+            }
+            catch (Throwable t) {
+                throw internalError(t);
+            }
+        }
+        if (hasNull) {
+            return null;
         }
+        return false;
+    }
 
-        for (int i = 0; i < leftPositionCount; i++) {
-            leftPositions[i] = i;
+    @TypeParameter("T")
+    @TypeParameterSpecialization(name = "T", nativeContainerType = boolean.class)
```

Comment:
TypeParameterSpecialization for boolean and double is not necessary. `boolean` is only used for boolean type and `double` is only used for `double`. But this is fine if you want consistency. Though there's too much code duplication. It would be nice to just have 5 separate functions for the `equals.invokeExact(xxx)` part and pass that method handle to a single implementation.